src/tests/authentication/test_token.py

```diff
@@ -0,0 +1,222 @@
+import json
+import uuid
+from datetime import datetime, timedelta
+from time import sleep
+from typing import Optional
+
+from django.contrib.auth import get_user_model
+from django.core.exceptions import ObjectDoesNotExist
+from django.test import TestCase
+
+from chaosinventory.authentication.models import Token, User
+
+
+def get_db_token(key) -> Token:
+    return Token.objects.get(key=key)
+
+
+def create_user(**kwargs) -> User:
+    password = None
+    if 'password' in kwargs:
+        password = kwargs.pop('password')
+
+    default_user_data = {
+        'username': uuid.uuid4(),
+        'first_name': 'Demo',
+        'last_name': 'User',
+        'email': 'user@example.org',
+    }
+    user: User = get_user_model().objects.create(**{**default_user_data, **kwargs})
+
+    if password is not None:
+        user.set_password(password)
+        user.save()
+
+    return user
+
+
+class TokenTestCase(TestCase):
+    def setUp(self) -> None:
+        self.user_password: str = 'password'
+        self.user = create_user(password=self.user_password)
+
+    def obtain_token(self, username: str = None, password: str = None, application: str = None,
+                     expiring: Optional[datetime] = None, renewable: bool = True):
+        payload = {
+            'username': self.user.username if username is None else username,
+            'password': self.user_password if password is None else password,
+            'application': self.__class__.__name__ if application is None else application,
+            'renewable': renewable,
+        }
+        if expiring is not None:
+            payload['expiring'] = expiring
+
+        return self.client.post(
+            '/api/authentication/token/credentials',
+            payload,
+        )
+
+    def get_token(self, response=None, **kwargs):
+        if response is None:
+            response = self.obtain_token(**kwargs)
+
+        return json.loads(response.content)['token']
+
+    def assertValidJson(self, content: str):
+        try:
+            return json.loads(content)
+        except json.JSONDecodeError:
+            self.fail('Failed to parse JSON')
```

Comment:
That section should probably also fail on other errors/exceptions?

Reply:
Generally yes, but in this case I want to explicitly validate if it's a json response. Other exceptions will also make the test fail / marked as error.